gms/gossiper.cc

```diff
@@ -1166,7 +1166,17 @@ version_type gossiper::get_max_endpoint_state_version(endpoint_state state) cons
 
 future<> gossiper::evict_from_membership(inet_address endpoint, permit_id pid) {
     verify_permit(endpoint, pid);
-    _unreachable_endpoints.erase(endpoint);
+    co_await mutate_live_and_unreachable_endpoints([endpoint] (gossiper& g) {
+        g._unreachable_endpoints.erase(endpoint);
+        g._unreachable_endpoints_version++;
+        // When evict_from_membership is called, the endpoint should have
+        // already been erased from _live_endpoints, so enforce that here.
+        auto was_live = g._live_endpoints.erase(endpoint);
+        if (was_live) {
+            on_internal_error_noexcept(logger, fmt::format("Node {} was unexpectedly marked as live when evicted from membership", endpoint));
```

Comment:
This is a more generic issue than node to be evicted should be not in live_endpoints. It is the issue that we need to make sure the _live_endpoints are in sync with _endpoint_state_map. 

Perhaps we can add some check in gossiper::run() to make sure _live_endpoints and _endpoint_state_map are in sync. Since now we reply much more on the _live_endpoints than before. 

I'd suggest drop the on_internal_error_noexcept because it is not a serious problem that deserve an abort, we are removing the node anyway.

Reply:
okay.  the consistency of endpoint_state with _live_endpoints is also checked (in this branch) in gossiper::is_alive in non-release modes.
It is not symmetric as we want to require the all node in _live_endpoints have an endpoint_state, but not vice verse.
It is possible to have an endpoint_state of a node that has left the ring and is not in _live_endpoints.